qa/rpc-tests/wallet_shieldcoinbase.py

```diff
@@ -71,44 +70,47 @@ def run_test (self):
         self.nodes[2].importaddress(mytaddr)
         try:
             self.nodes[2].z_shieldcoinbase(mytaddr, myzaddr)
-        except JSONRPCException,e:
-            errorString = e.error['message']
-        assert_equal("Could not find any coinbase funds to shield" in errorString, True)
+        except JSONRPCException as e:
+            errorString_01 = e.error['message']
+        assert_equal("Could not find any coinbase funds to shield" in errorString_01, True)
```

Comment:
It may be beyond the scope of this PR, but it would be nice if these try...except...assert blocks read as follows:
```
try:
    self.nodes[i].do_something()
    fail("An exception should have been thrown when trying to do something")
except JSONRPCException as e:
    assert_true("An error" in e.error['message'])
```
This would cover failing the test if the exceptions were not thrown, but prevent the errorStrings from having to exist outside the scope of the catch block.

Reply:
Are you familiar with `pytest`?   It has robust idioms for common cases like this.   Indeed for this particular "expected exception" type of test it has this syntax:

https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

Anywho, I agree that it's better to not leak an `errorString` variable to unnecessarily wide scopes.  (Note that, prior to my version of this code there was _one_ `errorString` that was _potentially_ bound to a new message at each expected "exception test".  I hacked in the unique name per expected exception e.g. `errorString_01` to prevent confusion about which value was bound.

Your solution is better.  I'll use it.   

In future I hope this kind of test suite can be handled with a proper test framework (i.e. pytest).

Reply:
I am collecting a list of improvements that can/should be implemented in the test code.

I'm not sure where to publish this list.   Should I open an issue about rpc test code quality?

Reply:
A proper testing framework would be nice.... If you made an issue with a list of potential improvements for the rpc tests, that would be great!

Reply:
The built-in `unittest` framework has `with self.assertRaises(Exception): ...`

Python 2: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertRaises
Python 3: https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises

Reply:
```
with self.assertRaises(JSONRPCException) as cm:
    self.nodes[i].do_something()

self.assertIn("An error", cm.exception.error['message'])
```
or possibly (untested; I'm not sure if this works for `JSONRPCException`):
```
with self.assertRaisesRegex(JSONRPCException, ".*An error.*"):
    self.nodes[i].do_something()
```

Reply:
Note that this wouldn't currently work because [BitcoinTestFramework doesn't subclass unittest.TestCase](https://github.com/zcash/zcash/blob/22dc3c04a419e27be914b529ee39a840b387ccea/qa/rpc-tests/test_framework/test_framework.py#L23). But that could be easily fixed, I think.

Reply:
No problem having `BitcoinTestFramework` inherit from `unittest.TestCase`.